### ISSUE TITLE
use population (not sample) standard deviation in computing zscores

### ISSUE THREAD

**Issue by maxheld83** [aiorazabala/qmethod#65]
in `qzscores.R`, the factor array zscores are currently computed using `sd()`, which, according to the R documentation:

> Like var this uses denominator n - 1.

That seems appropriate for *estimating* the population sd from a sample sd, but in this case, zscoring over all items, the *population* sd variant would seem to be more appropriate.

There appears (oddly) to be no population sd function in R, but it's [an easy 1-line fix](http://stackoverflow.com/questions/6457755/standard-deviation-in-r-seems-to-be-returning-the-wrong-answer-am-i-doing-some).

The bias introduced is probably miniscule, but still.

**Comment by maxheld83** [comment]
@aiorazabala let me know if I got this right and you agree; I'll send you a PR (with only that tiny fix) asap.

**Comment by aiorazabala** [comment]
Good point. This would require re-validating the results though (i.e. comparing the results of two-three samples, selecting 2 to 6 factors---i.e. about 10-15 analyses, against those results in PQMethod).

I'd say that we'd want to be 100% sure that this is correct, before implementing it, or either test it as above :S

**Comment by maxheld83** [comment]
yikes, of course. I hadn't thought about that.
I'll go through Brown and `pqmethod` and, if necessary, the email list to figure out what the thinking on this is.
We should definitely be very, very sure before opening this can of worms (validation) :smirk:

**Comment by maxheld83** [comment]
I've scoured Brown 1980, and did not find an explicit statement about using the population or sample standard deviation, but his example on page 241 indicates that *he* at least, did, in fact, use the **population** sd.

Here's an example R script to reproduce the analyses:

```
# this data is from Brown 1980: 241, just to test which SD he was using
factor_b <- c(
  58,
  2,
  -58,
  16,
  -24,
  0,
  -76,
  -47,
  39,
  -81,
  -42,
  7,
  32,
  -37,
  -20,
  -57,
  -2,
  21,
  39,
  -15,
  29,
  -28,
  -10,
  23,
  17,
  42,
  -36,
  1,
  -15,
  21,
  -14,
  41,
  9
)
sum(factor_b)  # gives -165, as in the book
mean(factor_b)  # gives -5, as in the book
sd(factor_b)  # gives something a litle larger than the book.
sd(factor_b) * sqrt((length(factor_b)-1)/length(factor_b))  # gives the book result
```

**Comment by maxheld83** [comment]
actually, what I said was wrong – Brown 1980: 241 does, in fact *explicitly* state (in formula) that he uses the **population sd**, the formula is, after all, `s`, conventionally stands for the *population* sd.
![z-formula-brown](https://cloud.githubusercontent.com/assets/5372770/6488498/37076e3a-c298-11e4-9bf2-e2faebd1a127.png)

**Comment by maxheld83** [comment]
I've just posted this on Cross Validated (the Stack Exchange for statistics): http://stats.stackexchange.com/questions/140328/standardizing-sample-factor-scores-population-or-standard-deviation

**Comment by maxheld83** [comment]
in another strike against sample sd, [wikipedia](http://en.wikipedia.org/wiki/Standard_score) defines zscores as using the *population sd*.

Think we'll have to raise this to level of bug, though it's important to recognize that any effect on results will likely be quite small (second-digit kind-of-small) – or am I wrong?

**Comment by maxheld83** [comment]
just wrote the below email to Peter Schmolck and Steven Brown, let's see what they have to say:

> Dear Dr Brown,
> Dear Peter,
>
>I have a small technical question about calculating zscores that arose during programming (on the margins of) Aiora's R package `qmethod`. I doubt it will be of interest to most people on the qmethod list, so I'm sending it just to you two. (If you think it is of interest to more people, feel free to forward it, of course).
>
> During calculating the standardized, weighted factor scores, would you recommend using the population (1/N) or the (corrected) sample standard deviation (1/n-1)?
>
> I am aware that Brown (1980: 241f) uses the population SD, and that standardized scores are also conventionally defined with reference to the population, not sample SD.
>
> Still, I wanted to make sure we get this right for the R package `qmethod`. I also had to think a little harder about this, because in Q methodology, after all, items are sampled from a broader concourse (not people, as in survey research), and therefore we are (sort of) standardizing over elements from a sample in this case.
> 
> Still, I came out in favor of the population standard deviation, because we're just interested in standardizing here, not some inference to a broader group (of items). Correct?
> 
> I was also wondering how this was implemented in `PQmethod` – would be good to know, so we can really fine-tune and validate `qmethod` on R.
> 
> Any thoughts on this, just a short "yes" or "no" would be greatly appreciated.
>
> Best
> Max
> 
> Ps.: This question also exists as an issue on our qmethod repository for R development, and on CrossValidated, a professional question-and-answer site for statisticians, though I suspect an email reply will be more convenient for you.

**Comment by maxheld83** [comment]
I just received a reply from Peter Schmolck

> Dear Maximilian,
> z-standardization has nothing to do with inference from sample to population. Its purpose is making scores comparable between variables or, here, betweeen q-factors such that all variables after z-standardization have the same mean and SD. In the z-scale the mean is set to 0, the SD to 1. Other possibilies for standard scores are, e.g. T-scale with mean 50 and SD 10. This version makes for simpler scores: no decimals necessary and only positive scores. Would you use the n-1 formula (for s instead of sigma) this would affect all factor scores the same way, because the n statements are the same.
>
> Peter

And from Steven Brown:

> Maximilian
>
> Both the SD divisor of N and also n-1 are used in the PQMethod, but I
think that no one knows this (perhaps not even Peter Schmolck!) and it is
so trivial as hardly to deserve mention.  There is a table in the PQMethod
output (on p. 2 of the attached) that lists the means and standard
deviations of all the Q sorts (SD = 1.864), and this is the outcome that
used n-1 in the divisor.  Later, at the bottom of the table of factor
arrays (p. 7), the standard deviation is listed as SD = 1.817, which used
N.  If I had to choose, I would vote for N in both instances, but these
statistics are not used for inferential purposes and in fact play
practically no meaningful role at all in the calculations.
>
> Best wishes,
>
> Steven

so:

1. it *should* be the population SD
2. it doesn't (much) matter, because all factors are affected in the same way
3. I really should've thought of 2), that's pretty stupid of me.

This doesn't qualify for "bug" anymore.

I suggest we fix this at some point, if we have automated testing (#89) in place, by which we can programmatically test whether `qmethod` still produces in a number of cases / datasets / etc. what other, published analyses produce.

**Comment by maxheld83** [comment]
I just added a small helper function to calculate the population sd.
It's just a trivial wrapper for `sd()`, but it saves some time typing.
I decided to also expose, document to the user, in case the user may wish to calculate some pop.sd manually (I find myself that doing a lot).
It's part of the `array-viz` branch and I'll PR it together with the other graphs once ready.

The helper function, of course, does not fix the (baby-)bug in `qzscores()` yet, nor accomplish the needed validation :/.

**Comment by maxheld83** [comment]
duplicateion https://github.com/QWrks/pensieve/issues/259
